cpp/tests/rolling/range_comparator_test.cu

```diff
@@ -0,0 +1,145 @@
+
+template <typename T>
+struct RangeComparatorTypedTest : RangeComparatorTest {};
+
+using TestTypes =
+  cudf::test::Concat<cudf::test::IntegralTypesNotBool, cudf::test::FloatingPointTypes>;
+
+TYPED_TEST_SUITE(RangeComparatorTypedTest, TestTypes);
+
+TYPED_TEST(RangeComparatorTypedTest, TestLessComparator)
+{
+  auto const less     = cudf::detail::nan_aware_less{};
+  auto constexpr nine = TypeParam{9};
+  auto constexpr ten  = TypeParam{10};
+
+  EXPECT_TRUE(less(nine, ten));
+  EXPECT_FALSE(less(ten, nine));
+  EXPECT_FALSE(less(ten, ten));
+
+  if constexpr (std::is_floating_point_v<TypeParam>) {
+    auto constexpr NaN = std::numeric_limits<TypeParam>::quiet_NaN();
+    auto constexpr Inf = std::numeric_limits<TypeParam>::infinity();
+    // NaN.
+    EXPECT_FALSE(less(NaN, ten));
+    EXPECT_FALSE(less(NaN, NaN));
+    EXPECT_FALSE(less(NaN, Inf));
+    EXPECT_FALSE(less(NaN, -Inf));
+    // Infinity.
+    EXPECT_TRUE(less(Inf, NaN));
+    EXPECT_FALSE(less(Inf, Inf));
+    EXPECT_FALSE(less(Inf, ten));
+    EXPECT_FALSE(less(Inf, -Inf));
+    // -Infinity.
+    EXPECT_TRUE(less(-Inf, NaN));
+    EXPECT_TRUE(less(-Inf, Inf));
+    EXPECT_TRUE(less(-Inf, ten));
+    EXPECT_FALSE(less(-Inf, -Inf));
+    // Finite.
+    EXPECT_TRUE(less(ten, NaN));
+    EXPECT_TRUE(less(ten, Inf));
+    EXPECT_FALSE(less(ten, -Inf));
+  }
+}
+
+TYPED_TEST(RangeComparatorTypedTest, TestGreaterComparator)
+{
+  auto const greater  = cudf::detail::nan_aware_greater{};
+  auto constexpr nine = TypeParam{9};
+  auto constexpr ten  = TypeParam{10};
+
+  EXPECT_FALSE(greater(nine, ten));
+  EXPECT_TRUE(greater(ten, nine));
+  EXPECT_FALSE(greater(ten, ten));
+
+  if constexpr (std::is_floating_point_v<TypeParam>) {
+    auto constexpr NaN = std::numeric_limits<TypeParam>::quiet_NaN();
+    auto constexpr Inf = std::numeric_limits<TypeParam>::infinity();
+    // NaN.
+    EXPECT_TRUE(greater(NaN, ten));
+    EXPECT_FALSE(greater(NaN, NaN));
+    EXPECT_TRUE(greater(NaN, Inf));
+    EXPECT_TRUE(greater(NaN, -Inf));
+    // Infinity.
+    EXPECT_FALSE(greater(Inf, NaN));
+    EXPECT_FALSE(greater(Inf, Inf));
+    EXPECT_TRUE(greater(Inf, ten));
+    EXPECT_TRUE(greater(Inf, -Inf));
+    // -Infinity.
+    EXPECT_FALSE(greater(-Inf, NaN));
+    EXPECT_FALSE(greater(-Inf, Inf));
+    EXPECT_FALSE(greater(-Inf, ten));
+    EXPECT_FALSE(greater(-Inf, -Inf));
+    // Finite.
+    EXPECT_FALSE(greater(ten, NaN));
+    EXPECT_FALSE(greater(ten, Inf));
+    EXPECT_TRUE(greater(ten, -Inf));
+  }
+}
+
+TYPED_TEST(RangeComparatorTypedTest, TestAddSafe)
+{
+  using T = TypeParam;
+  EXPECT_EQ(cudf::detail::add_safe(T{3}, T{4}), T{7});
+
+  if constexpr (cuda::std::numeric_limits<T>::is_signed) {
+    EXPECT_EQ(cudf::detail::add_safe(T{-3}, T{4}), T{1});
+  }
+
+  auto constexpr max = cuda::std::numeric_limits<T>::max();
+  EXPECT_EQ(cudf::detail::add_safe(T{max - 5}, T{4}), max - 1);
+  EXPECT_EQ(cudf::detail::add_safe(T{max - 3}, T{4}), max);
```

Comment:
I'd like to see an on the nose test here too:
```
EXPECT_EQ(cudf::detail::add_safe(T{max - 4}, T{4}, max);
```

Reply:
Good point. Fixed.